File: java/src/sorter/Witness.java
package sorter;

/**
 * This class will be utilized to verify the correctness of the records sorting algorithm
 */
public class Witness implements Provider {
    
    public final Provider source;

    // keeps track of all the records returned by next()
    private long count = 0;
    private long crc = 0;
    private Long lastKey = null;
    private boolean isSorted = true;
    
    public Witness(Provider source) {
        this.source = source;
    }

    /**
     * Returns the next record in the list and checks that each record key is sorted
     * @return next record or null if no more records exists
     */
    @Override
    public Record next() {
        Record result = source.next();
        if(result == null) return null;
        // checks that keys are sorted; sets isSorted to false if not
        if(isSorted) {
            long key = result.getKey();
            if(lastKey!=null) {
                if(lastKey>key) isSorted = false;
            }
            lastKey = key;
        }
        count++;
        // computes checksum value of record
        crc ^= result.getCrc();
        return result;
    }

    // returns the number of records returned by next()
    public long getCount() {
        return count;
    }

    // returns the checksum value
    public long getCrc() {
        return crc;
    }
    // returns whether the record key's/indexes have been sorted up to this point
    public boolean isSorted() {
        return isSorted;
    }

    @Override
    public String toString() {
        return "Witness [count=" + count + ", crc=" + crc + ", isSorted=" + isSorted + "]";
    }
    
    
}

File: Witness.cpp
#include "Witness.h"
#include "Providers/Provider.h"
/**
 * Initializes witness constructor and sets all values to their defaults
 * @param givenSource the source to get records from
 */
Witness::Witness(Provider* givenSource) {
    this->count = 0;
    this->crc = 0;
    this->isSorted = true;
    this->source = givenSource;
    this->lastRecord = nullptr;
}

/**
* Returns a pointer to the next record in the list and checks that each record key is sorted
* @return pointer to next record or a null pointer if no more records exists
*/

shared_ptr<Record> Witness::next() {
    // gets the next record pointer and returns the pointer if it is null
    shared_ptr<Record> recordPtr = source->next();
    if (recordPtr == nullptr) return nullptr;
    // else obtains the record and checks if it is sorted
    Record record = *recordPtr;
    if (isSorted){
        if (!lastRecord){
            if (record.compareTo(lastRecord) < 0)  isSorted = false;
        }
        this->lastRecord = recordPtr;
    }
    // increases count and computes the sequential check sum value
    count++;
    uint64_t checkSum = record.checksum();
    crc ^= checkSum;
    return recordPtr;
}

/**
  * Gets total number of generated records
  * @return the total number of record returned by next()
  */
long Witness::getCount() {
    return count;
}

/**
  * Returns the checksum value of all the records
  * @return checksum value
  */
uint64_t Witness::getCrc() {
    return crc;
}

/**
 * Returns a true or false value indicating whether the records are sorted or not
 * @return true if records are sorted else false
 */
bool Witness::checkSorted() {
    return isSorted;
}